File: 백준/Silver/2529. 부등호/부등호.py
n = int(input())
relations = list(input().split())

maxNum = float('-inf')
minNum = float('inf')
maxAnswer = ''
minAnswer = ''

def dfs(arr:list):
  global maxNum
  global minNum
  global maxAnswer
  global minAnswer

  if len(arr) == n+1:
    global isValid

    val = ''.join(list(map(str, arr)))
    isValid = True

    for i in range(len(relations)):
      if relations[i] == "<" and arr[i] > arr[i+1]:
        isValid = False
        break

      if relations[i] == '>' and arr[i] < arr[i+1]:
        isValid = False
        break
    
    if(isValid and int(val) > maxNum):
      maxNum = int(val)
      maxAnswer = val
    
    if(isValid and int(val) < minNum):
      minNum = int(val)
      minAnswer = val
    
    return
  
  for i in range(10):
    if i in arr:
      continue

    arr.append(i)
    dfs(arr)
    arr.pop()

dfs([])  

print(maxAnswer)
print(minAnswer)

File: 백준/Silver/2529. 부등호/부등호.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.replace(/\r/, ""));

// 0-9 까지 숫자가 단 하나만 들어갈 수 있다
// 해당 부등호를 만족하는 숫자조합을 최대 최소를 각각 출력해야된다
// step으로 바로 직전값과 해당 부등호를 보고 판단해서 넣으면된다.
const solution = (input) => {
  const n = Number(input[0]);
  const method = input[1].split(" ");
  const answer = [];

  const dfs = (arr, step) => {
    if (step === n) {
      answer.push(arr.join(""));
      return;
    }

    for (let i = 0; i <= 9; i++) {
      if (arr.includes(i)) continue;
      if (step === -1) {
        arr.push(i);
        dfs(arr, step + 1);
        arr.pop();
      } else {
        if (method[step] === "<" && arr[step] < i) {
          arr.push(i);
          dfs(arr, step + 1);
          arr.pop();
        }

        if (method[step] === ">" && arr[step] > i) {
          arr.push(i);
          dfs(arr, step + 1);
          arr.pop();
        }
      }
    }
  };

  dfs([], -1);

  console.log(answer[answer.length - 1]);
  console.log(answer[0]);
};

solution(input);
